cleanup-passes/formatter.ts

```diff
@@ -0,0 +1,164 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+'use strict';
+
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import {execFile} from 'child_process';
+import * as yaml from 'js-yaml';
+import * as semver from 'semver';
+import {promisify} from 'util';
+
+import {register} from '../cleanup-pass';
+import {ElementRepo} from '../element-repo';
+import {makeCommit} from './util';
+
+const execFilePromise = promisify(execFile);
+
+const formatterPackageName = 'webmat';
+const npmScriptName = 'format';
+const npmScriptCommand = 'webmat && npm run update-types';
+
+let latestFormatterVersion: string|undefined;
+async function getLatestFormatterVersion(): Promise<string> {
+  if (latestFormatterVersion === undefined) {
+    const {stdout} =
+        await execFilePromise('npm', ['info', formatterPackageName]);
+    const match = stdout.match(/latest: '(\d+\.\d+\.\d+)'/);
+    if (!match || !match[1]) {
+      throw new Error(
+          `Could not find latest version of ${formatterPackageName}`);
+    }
+    latestFormatterVersion = match[1];
+  }
+  return latestFormatterVersion;
+}
+
+/**
+ * This pass updates an element repo to format projects using
+ * https://github.com/PolymerLabs/webmat/ and configures a
+ * "format" NPM script that can be run to re-format the project.
+ *
+ * Throws an error if the repo has a missing or invalid package.json or if npm
+ * is not globally installed.
+ */
+async function formatPass(element: ElementRepo): Promise<void> {
+  const packageJsonPath = path.join(element.dir, 'package.json');
+  let packageJson: NpmConfig;
+  try {
+    packageJson = await fse.readJson(packageJsonPath);
+  } catch {
+    throw new Error(`${element.ghRepo.name}: Missing or invalid package.json.`);
+  }
+
+  // There are some changes we might write in this script that aren't important
+  // enough to make a commit for (e.g. just updating the package-lock.json).
+  let doCommit = false;
+
+  if (packageJson.devDependencies === undefined) {
+    packageJson.devDependencies = {};
+  }
+  if (packageJson.scripts === undefined) {
+    packageJson.scripts = {};
+  }
+
+  const newFormatterVersion = await getLatestFormatterVersion();
```

Comment:
This logic around checking version is probably kinda overkill, in the new pass (https://github.com/PolymerLabs/tedium/pull/75) I just hard coded the version number instead, since it only matters for major version bumps.